### ISSUE TITLE
Thread panic at block 14b054be76b515b12efcfdfd6e5d6bfe39988f937f48768c61e5c01600000000

### ISSUE THREAD

**Issue by rex4539** [paritytech/parity-zcash#50]
git @ b0ebed0

Reproducibility: always

Steps:
`pzec import /Zcash/blocks`

What happened:
```
...
2019-03-11 01:29:15  INFO sync Imported 230000 blocks
2019-03-11 01:30:40  INFO sync Imported 231000 blocks
2019-03-11 01:32:00  INFO sync Imported 232000 blocks
2019-03-11 01:33:44  INFO sync Imported 233000 blocks
2019-03-11 01:35:39  INFO sync Imported 234000 blocks
2019-03-11 01:37:24  INFO sync Imported 235000 blocks
thread 'main' panicked at 'Corrupted database - no sprout root for block 14b054be76b515b12efcfdfd6e5d6bfe39988f937f48768c61e5c01600000000', src/libcore/option.rs:1008:5
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:71
             at src/libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:211
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:227
   4: <std::panicking::begin_panic::PanicPayload<A> as core::panic::BoxMeUp>::get
             at src/libstd/panicking.rs:491
   5: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:398
   6: std::panicking::try::do_call
             at src/libstd/panicking.rs:325
   7: core::ptr::real_drop_in_place
             at src/libcore/panicking.rs:95
   8: <T as core::any::Any>::get_type_id
             at src/libcore/option.rs:1008
Abort trap: 6
```

Expected result:
All blocks are imported.

**Comment by svyatonik** [comment]
Are you importing to new database? The format has changed few times in last commits. This could possibly be a reason

**Comment by rex4539** [comment]
I'm trying to import the blocks synced with `zcashd`.

My expectation by issuing this command is that the blocks will be imported and used by `pzec`.

**Comment by svyatonik** [comment]
Your expectations are correct :) I mean - have there been some old `pzec` database before you started import? It could have been some (incompatible) database created by previous runs of `pzec`.

Anyway - could you please delete the `pzec` database (`rm -rf ~/.local/share/pzec/db` if you're on *nix) and retry import?

**Comment by rex4539** [comment]
Sure, I will try it and report back.

**Comment by rex4539** [comment]
Still stuck at the same point.

```
2019-03-28 18:24:44  INFO sync Imported 231000 blocks
2019-03-28 19:01:13  INFO sync Imported 232000 blocks
2019-03-28 20:08:33  INFO sync Imported 233000 blocks
2019-03-28 21:11:27  INFO sync Imported 234000 blocks
2019-03-28 22:13:16  INFO sync Imported 235000 blocks
Cannot append block
```

**Comment by svyatonik** [comment]
So I've finally just passed the range `235000...236000`:
```
2019-04-09 14:25:53  INFO sync Imported 227000 blocks
2019-04-09 14:29:20  INFO sync Imported 228000 blocks
2019-04-09 14:32:52  INFO sync Imported 229000 blocks
2019-04-09 14:38:07  INFO sync Imported 230000 blocks
2019-04-09 14:42:28  INFO sync Imported 231000 blocks
2019-04-09 14:44:49  INFO sync Imported 232000 blocks
2019-04-09 14:48:37  INFO sync Imported 233000 blocks
2019-04-09 14:58:31  INFO sync Imported 234000 blocks
2019-04-09 15:04:37  INFO sync Imported 235000 blocks
2019-04-09 15:10:57  INFO sync Imported 236000 blocks
2019-04-09 15:16:22  INFO sync Imported 237000 blocks
```

Still no idea what has caused your issue (could it be out of disk space?). I'll leave the issue open here for a while - probably someone I'll be able to repeat it or we'll have new ideas about it. What we actually can improve is reporting actual error instead of simple "Cannot append block" - I'll open PR in a moment.

**Comment by rex4539** [comment]
This issue occurs on both of my machines. They have plenty of disk space.

**Comment by rex4539** [comment]
```
2019-04-10 02:41:50  INFO sync Imported 235000 blocks
Cannot append block: Database(CannotCanonize)
```

**Comment by svyatonik** [comment]
Okay, think I have an idea. It could have been reorganization near that block (#235495) and your zcashd db contains competing block (not sure, but probably because you were syncing zcashd during that reorg), while mine does not (I've synced it yesterday). I'll prepare a fix today. Thanks for helping, @rex4539 !

**Comment by svyatonik** [comment]
@rex4539 This should be fixed by #78 .
